### DISCUSSION TITLE
How to write the type of an async pipe function?

### DISCUSSION THREAD

**Discussion by joshuaavalon** [millsp/ts-toolbelt#196] [Wiki]
#### Describe your question

How to write the type of an async pipe function?

**Implementation**
```typescript
const pipePromises = (...fns :any) => (x:any) => fns.reduce((p:any, fn:any) => p.then(fn), Promise.resolve(x));
const a = (a1: number) => `${a1}`;
const b = async (a1: string) => [a1];
pipePromises(a, b)(1).then((v: any) => console.log(v)); // ["1"]
```

`pipePromises` accepts normal function and async function. I know that the normal pipe function can use ` <Fns extends F.Function[]>(...args: F.Piper<Fns>): F.Pipe<Fns>` but I don't know how to write the promise version.

#### Search tags, topics

#promise, #async

**Comment by millsp** [comment]
td;dr It's now in the [docs](https://pirix-gh.github.io/ts-toolbelt/modules/_function_pipe_.html#pipe)




Hi @joshuaavalon,

I think it handles it by default:

```ts
const a = (a1: number) => `${a1}`;
const b = async (a1: string) => [a1];

const pipePromises = <Fns extends F.Function[]>(...fns: F.Piper<Fns>): F.Pipe<Fns> => {
  return ((x: any) => fns.reduce((p: any, fn: any) => p.then(fn), Promise.resolve(x))) as F.Pipe<Fns>
}

const t = pipePromises(a, b)(1).then((v: any) => console.log(v)); // ["1"]

```

or even

```ts
const a = async (a1: number) => `${a1}`;
const b = async (a1: Promise<string>) => [a1];

const pipePromises = <Fns extends F.Function[]>(...fns: F.Piper<Fns>): F.Pipe<Fns> => {
  return ((x: any) => fns.reduce((p: any, fn: any) => p.then(fn), Promise.resolve(x))) as F.Pipe<Fns>
}

const t = pipePromises(a, b)(1).then((v: any) => console.log(v)); // ["1"]
```

**Comment by millsp** [comment]
Use `as` to force a type cast

**Comment by joshuaavalon** [comment]
@pirix-gh I may failed to mention something in my example. The implementation actually unwrap the promise of the result.

Consider the following

```typescript
const a = async (a1: number) => `${a1}`;
const b = async (a1: string) => [a1];
``` 

This works because JavaScript automatically unwrap promise in `then`. `const b = async (a1: Promise<string>) => [a1];` is actually wrong because it will receive `string` instead of `Promise<string>`.

**Comment by millsp** [comment]
Yhea, I see what you mean now:

```ts
import {O, T, F} from 'ts-toolbelt';

const a = async (a1: number) => `${a1}`;
const b = async (b1: Promise<string>) => [b1];
const c = async (c1: Promise<Promise<string>[]>) => [c1];

const pipePromises = <Fns extends F.Function[]>(...fns: F.Piper<Fns>): F.Pipe<Fns> => {
  return ((x: any) => fns.reduce((p: any, fn: any) => p.then(fn), Promise.resolve(x))) as F.Pipe<Fns>
}

const t = pipePromises(a, b, c)(1).then((v: any) => console.log(v)); // ["1"]
```
However, this is still correct, but yours unwraps the value as it reduces the list.

I'll see what I can think of. I think you'll need a specially crafted type for this.

**Comment by millsp** [comment]
@joshuaavalon, good news!

You now have an async option that will unwrap the promise types.

Check the docs out. And let me know if it's all good :)

**Comment by millsp** [comment]
@joshuaavalon did it work out for you?
